pkg/watcher/reconciler/taskrun/taskrun_test.go

```diff
@@ -184,3 +188,100 @@ func TestDisableCRDUpdate(t *testing.T) {
 		t.Errorf("Did not expect change in TaskRun (-want, +got):\n%s", diff)
 	}
 }
+
+func TestRunCleanup(t *testing.T) {
+	for _, tc := range []struct {
+		gracePeriod time.Duration
+		wantDelete  bool
+	}{
+		{
+			gracePeriod: -1 * time.Second,
+			wantDelete:  true,
+		},
+		{
+			gracePeriod: 0,
+			wantDelete:  false,
+		},
+		{
+			gracePeriod: 1 * time.Second,
+			wantDelete:  true,
+		},
+	} {
+		t.Run(fmt.Sprintf("GracePeriod_%v", tc.gracePeriod), func(t *testing.T) {
+			env := newEnv(t, &reconciler.Config{
+				CompletedResourceGracePeriod: tc.gracePeriod,
+			})
+			fakeClock := clockwork.NewFakeClockAt(time.Now())
+			clock = fakeClock
+
+			tr, err := env.pipeline.TektonV1beta1().TaskRuns("ns").Create(&v1beta1.TaskRun{
+				TypeMeta: metav1.TypeMeta{
+					APIVersion: "v1beta1",
+					Kind:       "taskrun",
+				},
+				ObjectMeta: metav1.ObjectMeta{
+					Name:        "taskrun",
+					Namespace:   "ns",
+					Annotations: map[string]string{"demo": "demo"},
+					UID:         "12345",
+					OwnerReferences: []metav1.OwnerReference{{
+						APIVersion: "pipeline.tekton.dev/v1beta1",
+						Kind:       "PipelineRun",
+						Name:       "parent",
+						UID:        "asdf",
+					}},
+				},
+				Status: v1beta1.TaskRunStatus{
+					Status: duckv1beta1.Status{
+						Conditions: duckv1beta1.Conditions{
+							apis.Condition{
+								Type:   apis.ConditionSucceeded,
+								Status: corev1.ConditionTrue,
+							},
+						},
+					},
+					TaskRunStatusFields: v1beta1.TaskRunStatusFields{},
+				},
+			})
+			if err != nil {
+				t.Fatalf("Create: %v", err)
+			}
+
+			t.Run("noop-OwnerReference", func(t *testing.T) {
+				if err := env.ctrl.Reconciler.Reconcile(env.ctx, tr.GetNamespacedName().String()); err != nil {
+					t.Fatalf("Reconcile: %v", err)
+				}
+
+				// First run should be a no-op because of the OwnerReference.
+				if _, err := env.pipeline.TektonV1beta1().TaskRuns(tr.GetNamespace()).Get(tr.GetName(), metav1.GetOptions{}); err != nil {
+					t.Fatalf("Get(%s): %v", tr.Name, err)
+				}
+			})
+
+			t.Run("delete", func(t *testing.T) {
+				// Clear out OwnerRefs to make the object eligible for deletion.
+				tr.OwnerReferences = nil
+				if _, err := env.pipeline.TektonV1beta1().TaskRuns(tr.GetNamespace()).Update(tr); err != nil {
+					t.Fatalf("Update: %v", err)
+				}
+				// Force the wall clock forward (with some extra buffer) to
+				// simulate the progression of time past the configured grace
+				// period.
+				fakeClock.Advance(tc.gracePeriod + time.Minute)
```

Comment:
I'm not clear why the extra buffer is required - did you experience the test racing before this addition?

Reply:
FWIW not a major concern. But wondering if a buffer size of a minute might have potential to mask future timing bugs that could otherwise be caught?

Reply:
Yeah, it's because we calculate the grace period w.r.t. the Record's last update time, and since this package is external to the Result API, we don't have the same control over the API server's internal wall clock that we do in the API Server unit tests to fake things out to be deterministic.

Because we're mixing real clocks and fake clocks, we need to make sure that we're not stuck in the past when we try and compute the grace period during reconcile.
Instead this takes advantage of our control of the watcher's internal wall clock, fast forwarding to a time in the future that we can guarantee is 1) not before the record last update time and 2) past the grace period for grace period values > 0.

I think you're right that this still has a risk of racing if the time taken until this point is more than a minute (unlikely, but possible), so I added an extra bit to make sure the advancement is w.r.t. the current wall clock - it's now basically a sleep that happens instantaneously. 
As an alternative we could fast forward the clock w.r.t. the Record timestamp, but this probably isn't necessary since all the operations here are synchronous, so time.Now() works as an approximation.

tl;dr we're time traveling 🧙‍♂️